src/mapKeys.test.ts

```diff
@@ -105,4 +105,14 @@ describe("typing", () => {
     const result = mapKeys({ a: "b" }, constant(123));
     expectTypeOf(result).toEqualTypeOf<Partial<Record<123, string>>>();
   });
+
+  test("union of records", () => {
+    const data = {} as Record<number, string> | Record<string, string>;
+
+    const dataFirst = mapKeys(data, constant("hello" as string));
+    expectTypeOf(dataFirst).toEqualTypeOf<Record<string, string>>();
+
+    const dataLast = pipe(data, mapKeys(constant("hello" as string)));
+    expectTypeOf(dataLast).toEqualTypeOf<Record<string, string>>();
+  });
```

Comment:
I don't think this test tests what this PR fixes. Because you use constant for the new key, there's nothing testing the result of the union. You should test different value types (and the key doesn't matter): `Record<PropertyKey, "cat"> | Record<PropertyKey, "dog">`, or test the mapper itself like you test in forEach.